contracts/JumpBall.cdc

```diff
@@ -0,0 +1,379 @@
+
+            let creatorTotal = self.calculateTotal(game: game, address: game.creator, stats: stats)
+            let opponentTotal = self.calculateTotal(game: game, address: game.opponent, stats: stats)
+
+            if creatorTotal > opponentTotal {
+                self.awardWinner(game: game, winner: game.creator, winnerCap: game.getCreatorCap())
+            } else if opponentTotal > creatorTotal {
+                let opponentCap = game.getOpponentCap() ?? panic("Opponent capability not found.")
+                self.awardWinner(game: game, winner: game.opponent!, winnerCap: opponentCap)
+            } else {
+                game.returnAllNFTs()
+            }
+        }
+
+        access(contract) fun calculateTotal(game: &Game, address: Address?, stats: {UInt64: UInt64}): UInt64 {
+            if address == nil {
+                return 0
+            }
+
+            var total: UInt64 = 0
+            let nftKeys = game.getNFTKeys()
+
+            for key in nftKeys {
+                let owner = game.getOwnership(key: key) ?? panic("Owner not found.")
+                if owner == address {
+                    total = total + (stats[key] ?? 0)
+                }
+            }
+
+            return total
+        }
+
+        access(contract) fun awardWinner(game: &Game, winner: Address, winnerCap: Capability<&{NonFungibleToken.Collection}>) {
+            emit WinnerDetermined(gameID: game.id, winner: winner)
+            game.transferAllToWinner(winner: winner, winnerCap: winnerCap)
+        }
+    }
+
+    // Factory function to create a new Player resource
+    access(all) fun createPlayer(playerAddress: Address, collectionCap: Capability<&{NonFungibleToken.Collection}>): @Player {
+        if !collectionCap.check() {
+            panic("A valid NFT collection capability is required to create a player.")
+        }
+
+        return <- create Player(
+            address: playerAddress,
+            collectionCap: collectionCap
+        )
+    }
+
+    access(all) var metadata: {String: AnyStruct}
+    access(all) var games: @{String: Game}
+    access(all) var userGames: {Address: [String]}
+    access(all) let admin: @Admin
+
+    init() {
+        self.metadata = {}
+        self.admin <- create Admin()
+        self.games <- {}
+        self.userGames = {}
+
+        // Initialize Player resource for contract owner
+        let collectionCap = self.account.capabilities.storage.issue<&{NonFungibleToken.Collection}>(/storage/OwnerNFTCollection)
+        self.account.capabilities.publish(collectionCap, at: /public/OwnerNFTCollection)
+
+        // Save the Player resource in storage
+        self.account.storage.save<@Player>(<- create Player(
+            address: self.account.address,
+            collectionCap: collectionCap
+        ), to: /storage/JumpBallPlayer)
+    }
+
+    // Helper functions for contract metadata
+    access(all) fun setMetadata(key: String, value: AnyStruct) {
```

Comment:
This can be called by anyone - we can address by making it access(contract) and wrapping in admin resource, or we just don't have a setMetadata method if we are not currently making use of the metadata map